rust/kernel/time.rs

```diff
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! This module contains timing methods for functionality like sleeping.
+
+use core::ffi::c_uint;
+
+use crate::bindings::msleep as msleep_bindgen;
+
+/// This method sleeps the current thread for `msecs` milliseconds, calling the
+/// kernel method msleep.
+pub fn msleep(msecs: c_uint) {
```

Comment:
This should be `core::time::Duration`.

One difficulty here is that we can't represent all durations with a `c_uint`. I think we need to write a converter function that saturates the `c_uint` if the duration is greater than `UINT_MAX` and perhaps rounds up to 1 if the duration is less than 1ms (but greater than zero).

Reply:
Would it be acceptable to round up in other cases where there's a non-zero nano or microsecond value? I feel like that is more consistent if we're already rounding up in the less than 1 millisecond case.

Reply:
I am skeptical whether `core::time::Duration` is the best abstraction for us. Many kernel APIs use microsecond precision, and `Duration` is a mapping of timespec (i.e. second + subsecond nanos); conversion from `Duration` to a microsecond value will require division and multiplication and thus not exactly cheap. We probably want some types that encode units, e.g. `Nanoseconds`, `Microseconds`. `Milliseconds`, `Seconds` and implement `From` traits so that they can interconvert and convert from/to `Duration`.

Reply:
The `Duration` logic does get a bit messy with the suggested approach. I need to think about it a little bit more, but this at least requires one multiplication, one division, and one mod operation.

We need to get the duration in nanoseconds to see if we round up to 1 millisecond. Because nanoseconds is encoded as a `u128`, we can then do one modulo operation to see if we need to round up (zero stays zero, nonzero rounds up to 1 millsecond). From there, `u128::MAX / Duration::MILLISECOND.as_nanos() > u64::MAX` so we could assume that for the `unsigned int` input of `msleep` (even if there's an architecture where `unsigned int` is a typedef for `uint64_t`), the result of dividing nanoseconds into milliseconds will not cause any loss of representation of milliseconds because the maximum representation of `unsigned int` will still be less than `u128::MAX / Duration::MILLISECOND.as_nanos()`.

I'm happy to try either way, but the `Milliseconds` approach may turn out to be cleaner in cases where an integer is required as input on the C side.

Reply:
I think my other thought is that I could see using `Duration` being confusing to users unless we very explicitly document the rounding/saturating behavior of `msleep`.

Reply:
`msleep` already has divisions and rounding up to convert from ms to jiffies.

Anyway, I'm fine with new types to represent durations and reduce the cost of conversion when it isn't needed. What I'd like to avoid is naked types (e.g., u32, u64) whose units aren't immediately obvious.

Reply:
That makes sense! I'll modify this PR to be consistent with @nbdd0121's suggestion.